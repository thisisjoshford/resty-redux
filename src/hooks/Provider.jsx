import React, { createContext, useState, useContext } from 'react';
import { fetchRequest } from '../services/fetchRequest';

const RestyContext = createContext();

export const RestyProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('GET');
  const [textBody, setTextBody] = useState('');
  const [response, setResponse] = useState([]);
  const [history, setHistory] = useState([]);

  const onChange = ({ target }) => {
    if(target.name === 'url') setUrl(target.value);
    if(target.name === 'method') setMethod(target.value);
    if(target.name === 'textBody') setTextBody(target.value);
  };

  const onSubmit = event => {
    event.preventDefault();
    fetchRequest(url, method, textBody)
      .then(res => {
        setResponse(res);
        setHistory(prevHistory => 
          [{ url, method }, ...prevHistory]);
      });
  };

  return (
    <RestyContext.Provider 
      value ={{ 
        url, 
        method, 
        textBody, 
        response,
        history, 
        onChange, 
        onSubmit 
      }}>
      {children}
    </RestyContext.Provider>
  );
};

export const useProvider = () => {
  const context = useContext(RestyContext);
  return context;
};
