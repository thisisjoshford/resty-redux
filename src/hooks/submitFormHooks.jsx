import { fetchRequest } from '../services/fetchRequest';
import { useDispatch, useSelector } from 'react-redux';
import { getUrl, getMethod, getTextBody, getResponse, getHistory } from '../selectors/selectors';
import { addResponse, addHistory, setUrl, setMethod, setTextBody } from '../actions/restfulActions';

export const useSubmitForm = () => {
  const dispatch = useDispatch();
  const url = useSelector(getUrl);
  const method = useSelector(getMethod);
  const textBody = useSelector(getTextBody);
  const response = useSelector(getResponse);
  const history = useSelector (getHistory);

  const onChange = ({ target }) => {
    if(target.name === 'url') dispatch(setUrl(target.value));
    if(target.name === 'method') dispatch(setMethod(target.value));
    if(target.name === 'textBody') dispatch(setTextBody(target.value));
  };

  const onSubmit = event => {
    event.preventDefault();
    dispatch(fetchRequest(url, method, textBody)
      .then(res => {
        addResponse(res);
        addHistory(prevHistory => 
          [{ url, method }, ...prevHistory]);
      }));
  };

  return {
    url,
    method,
    textBody,
    response,
    history,
    onChange,
    onSubmit
  };
};
