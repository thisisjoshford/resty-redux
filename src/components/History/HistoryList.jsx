import React from 'react';
import HistoryItem from './HistoryItem';
import { useProvider } from '../../hooks/Provider';

const HistoryList = () => {
  const { history } = useProvider();
  const historyNodes = history
    .map((item, i) => (
      <li key={i}>
        <HistoryItem {...item} />
      </li>
    ));
  
  return (
    <ul>
      {historyNodes}
    </ul>
  );
};

export default HistoryList;
