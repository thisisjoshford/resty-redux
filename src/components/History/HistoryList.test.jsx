import React from 'react';
import { shallow } from 'enzyme';
import HistoryList from './HistoryList';
import { RestyProvider } from '../../hooks/Provider';

describe('History list component', () => {
  it('renders History list', () => {

    const wrapper = shallow(
      <RestyProvider>
        <HistoryList/>
      </RestyProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
