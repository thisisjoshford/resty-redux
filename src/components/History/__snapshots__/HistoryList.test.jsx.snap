// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History list component renders History list 1`] = `
<ContextProvider
  value={
    Object {
      "history": Array [],
      "method": "GET",
      "onChange": [Function],
      "onSubmit": [Function],
      "response": Array [],
      "textBody": "",
      "url": "",
    }
  }
>
  <HistoryList />
</ContextProvider>
`;
